app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java

```diff
@@ -136,125 +156,297 @@ public void setStudyIdIndex(DynamoIndexHelper studyIdIndex) {
-        }
+    // Helper method to convert a survey question into one or more schema field defs. As some survey questions can
+    // generate more than one field (such as unit fields), callers should pass in a field def list, and this method
+    // will append created field defs to that list.
+    private static void addFieldDefsForSurveyQuestion(List<UploadFieldDefinition> fieldDefList,
+            SurveyQuestion question) {
+        // These preconditions should never happen, but we have a Preconditions check here just in case.
+        Preconditions.checkNotNull(question);
+
+        Constraints constraints = question.getConstraints();
+        Preconditions.checkNotNull(constraints);
 
+        DataType surveyQuestionType = constraints.getDataType();
+        Preconditions.checkNotNull(surveyQuestionType);
+
+        // Init field def builder with basic fields. Note that all survey questions are skippable, so mark the field as
+        // optional (not required).
+        String fieldName = question.getIdentifier();
+        DynamoUploadFieldDefinition.Builder fieldDefBuilder = new DynamoUploadFieldDefinition.Builder()
+                .withName(fieldName).withRequired(false);
+
+        UploadFieldType uploadFieldType;
         if (constraints instanceof MultiValueConstraints) {
             MultiValueConstraints multiValueConstraints = (MultiValueConstraints) constraints;
-
             if (multiValueConstraints.getAllowMultiple()) {
-                if (constraints.getDataType() == DataType.STRING) {
-                    // Multiple choice string answers frequently become longer than the 100-char limit. This will have
-                    // to be a JSON attachment.
-                    return UploadFieldType.ATTACHMENT_JSON_BLOB;
-                } else if (multiValueConstraints.getEnumeration() != null
-                        && multiValueConstraints.getEnumeration().size() > MAX_MULTI_VALUE_INT_LENGTH) {
-                    // Even if it's not a string, if there are too many options and someone selects all of them, it
-                    // could exceed the 100-char limit.
-                    return UploadFieldType.ATTACHMENT_JSON_BLOB;
-                } else {
-                    // Multiple choice non-string answers (frequently ints) tend to be very short, so this can fit in
-                    // an inline_json_blob.
-                    return UploadFieldType.INLINE_JSON_BLOB;
+                uploadFieldType = UploadFieldType.MULTI_CHOICE;
+                fieldDefBuilder.withAllowOtherChoices(multiValueConstraints.getAllowOther());
+
+                // convert the survey answer option list to a list of possible multi-choice answers
+                List<String> fieldAnswerList = new ArrayList<>();
+                //noinspection Convert2streamapi
+                for (SurveyQuestionOption oneSurveyOption : multiValueConstraints.getEnumeration()) {
+                    fieldAnswerList.add(oneSurveyOption.getValue());
```

Comment:
Value (server value) can be null. When it's null, the survey engine is supposed to return the label for the option. There are definitely surveys in production that do this.